lisa/tests/staging/scheduler/util_tracking.py

```diff
@@ -0,0 +1,377 @@
+                         (df.next_state == TaskState.TASK_ACTIVE)].index
+
+        # Check task signals at each activation
+        df = self.trace.df_events('sched_util_est_task')
+        df = df[df.pid == test_task.pid]
+
+        # Define a time interval to correlate relative trace events.
+        def restrict(df, time, delta=1e-3):
+            return df[time-delta:time+delta]
+
+        failures = []
+        passed = True
+        for idx, activation in enumerate(activations):
+            avg, enq, ewma = restrict(df, activation)[[
+                'util_avg', 'util_est_enqueued', 'util_est_ewma']].iloc[-1]
+
+            metrics[idx+1] = ActivationSignals(activation, avg, enq, ewma)
+
+            # UtilEst is not updated when within 1% of previous activation
+            if 1.01 * enq < avg:
+                failure_reasons[idx] = 'enqueued({}) smaller than util_avg({}) @ {}'\
+                                        .format(enq, avg, activation)
+                failures.append(activation)
+                passed = False
+                continue
+
+            # Running on FastRamp kernels:
+            if self.fast_ramp:
+
+                # STABLE, DOWN and UP:
+                if enq > ewma:
+                    failure_reasons[idx] = 'enqueued({}) bigger than ewma({}) @ {}'\
+                                            .format(enq, ewma, activation)
+                    failures.append(activation)
+                    passed = False
+
+            # Running on (legacy) non FastRamp kernels:
+            else:
+
+                phase = self.trace.analysis.rta.task_phase_at(test_task, activation)
+
+                # STABLE: ewma ramping up
+                if phase.id == 0 and enq < ewma:
+                    failure_reasons[idx] = 'enqueued({}) smaller than ewma({}) @ {}'\
+                                            .format(enq, ewma, activation)
+                    failures.append(activation)
+                    passed = False
+                    continue
+
+                # DOWN: ewma ramping down
+                if 0 < phase.id < 5 and enq > ewma:
+                    failure_reasons[idx] = 'enqueued({}) bigger than ewma({}) @ {}'\
+                                            .format(enq, ewma, activation)
+                    failures.append(activation)
+                    passed = False
+                    continue
+
+                # UP: ewma ramping up
+                if phase.id > 4 and enq < ewma:
+                    failure_reasons[idx] = 'enqueued({}) smaller than ewma({}) @ {}'\
+                                            .format(enq, ewma, activation)
+                    failures.append(activation)
+                    passed = False
+                    continue
+
+        bundle = ResultBundle.from_bool(passed)
+        bundle.add_metric("Signals", metrics)
+        if not passed:
+            bundle.add_metric("Reasons", failure_reasons)
```

Comment:
s/Reasons/failure reasons/ to make explicit the reason of what we are talking about